File: solutions/java/src/librarymanagementsystem/LibraryManagementSystemDemo.java
package librarymanagementsystem;

import java.util.List;

public class LibraryManagementSystemDemo {
    public static void run() {
        LibraryManagementSystem libraryManagementSystem = LibraryManagementSystem.getInstance();

        // Add books to the catalog
        BookCopy copy1 = libraryManagementSystem.addBookCopy("ISBN1", "Book 1", "Author 1");
        BookCopy copy2 = libraryManagementSystem.addBookCopy("ISBN2", "Book 2", "Author 1");
        BookCopy copy3 = libraryManagementSystem.addBookCopy("ISBN3", "Book 3", "Author 3");

        // Register members
        Member member1 = libraryManagementSystem.registerMember("John Doe", "john@example.com");
        Member member2 = libraryManagementSystem.registerMember("Jane Smith", "jane@example.com");

        // Borrow books
        libraryManagementSystem.borrowBook(member1.getId(), copy1.getBarCode());
        libraryManagementSystem.borrowBook(member2.getId(), copy2.getBarCode());

        // Return books
        libraryManagementSystem.returnBook(member1.getId(), copy1.getBarCode());

        // Search books
        List<Book> searchResults = libraryManagementSystem.searchByAuthor("Author 1");
        System.out.println("Search Results for Author1:");
        for (Book book : searchResults) {
            System.out.println(book.getTitle() + " by " + book.getAuthor());
        }
    }
}

File: solutions/golang/librarymanagementsystem/library_management_system_demo.go
package librarymanagementsystem

import (
	"fmt"
)

func Run() {
	libraryManager := GetLibraryManager()

	// Add books to the catalog
	book1 := NewBook("ISBN1", "Book 1", "Author 1", 2020)
	book2 := NewBook("ISBN2", "Book 2", "Author 2", 2019)
	book3 := NewBook("ISBN3", "Book 3", "Author 3", 2021)

	libraryManager.AddBook(book1)
	libraryManager.AddBook(book2)
	libraryManager.AddBook(book3)

	// Register members
	member1 := NewMember("M1", "John Doe", "john@example.com")
	member2 := NewMember("M2", "Jane Smith", "jane@example.com")

	libraryManager.RegisterMember(member1)
	libraryManager.RegisterMember(member2)

	// Borrow books
	if err := libraryManager.BorrowBook("M1", "ISBN1"); err != nil {
		fmt.Printf("Error borrowing book: %v\n", err)
	}

	if err := libraryManager.BorrowBook("M2", "ISBN2"); err != nil {
		fmt.Printf("Error borrowing book: %v\n", err)
	}

	// Return books
	if err := libraryManager.ReturnBook("M1", "ISBN1"); err != nil {
		fmt.Printf("Error returning book: %v\n", err)
	}

	// Search books
	searchResults := libraryManager.SearchBooks("Book")
	fmt.Println("\nSearch Results:")
	for _, book := range searchResults {
		fmt.Printf("%s by %s\n", book.Title, book.Author)
	}
}
